File: src/main/java/ffblockly/meta/AvOptionBuilder.java
package ffblockly.meta;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;

public class AvOptionBuilder {
    private Matcher optMatcher;

    public AvOptionBuilder setOptMatcher(Matcher optMatcher) {
        this.optMatcher = optMatcher;
        return this;
    }

    public AvOption  createAvOptionImpl() {
        final Matcher optMatcher1 = optMatcher;
        return Scan.MYFACTORY.options(new AvOption() {
            private final Matcher optMatcher = optMatcher1;
            protected List<AvOption> ch;
            private List<AvTarget> r;

            @Override
            public String getName() {
                return this.optMatcher.group(AvOptionField.name.name());
            }

            @Override
            public String getType() {
                return this.optMatcher.group(AvOptionField.type.name());
            }

            @Override
            public List<AvTarget> getTargets() {
                if (this.getType() != null)
                    if (null == this.r) {
                        this.r = new ArrayList<>();
                        String group = this.optMatcher.group(AvOptionField.targets.name());
                        for (AvTarget avTarget : AvTarget.values())
                            if ('.' != group.charAt(avTarget.ordinal()))
                                this.r.add(avTarget);
                    }
                return this.r;
            }

            @Override
            public String getDescription() {
                return this.optMatcher.group(AvOptionField.description.name());
            }

            @Override
            public String getRange1() {
                return this.optMatcher.group(AvOptionField.range1.name());
            }

            @Override
            public String getRange2() {
                return this.optMatcher.group(AvOptionField.range2.name());
            }

            @Override
            public String getDef() {
                return this.optMatcher.group(AvOptionField.def.name());
            }

            @Override
            public Boolean getEncoding() {
                return null != this.getType() ? this.getTargets().contains(AvTarget.Encoding) ? true : null : null;
            }

            @Override
            public Boolean getDecoding() {
                return null != this.getType() ? this.getTargets().contains(AvTarget.Decoding) ? true : null : null;
            }

            @Override
            public Boolean getFiltering() {
                return null != this.getType() ? this.getTargets().contains(AvTarget.Filtering) ? true : null : null;

            }

            @Override
            public Boolean getVideo() {
                return null != this.getType() ? this.getTargets().contains(AvTarget.Video) ? true : null : null;

            }

            @Override
            public Boolean getAudio() {
                return null != this.getType() ? this.getTargets().contains(AvTarget.Audio) ? true : null : null;

            }

            @Override
            public Boolean getSubtitle() {
                return null != this.getType() ? this.getTargets().contains(AvTarget.Subtitle) ? true : null : null;

            }

            @Override
            public Boolean getExport() {
                return null != this.getType() ? this.getTargets().contains(AvTarget.Export) ? true : null : null;

            }

            @Override
            public Boolean getReadonly() {
                return null != this.getType() ? this.getTargets().contains(AvTarget.Readonly) ? true : null : null;
            }

            @Override
            public List<AvOption> getChildren() {
                return this.ch;
            }

            @Override
            public void addChild(AvOption o) {
                if (this.ch == null)
                    this.ch = new ArrayList<>();
                this.ch.add(o);
            }
        }).as();
    }
}

File: src/main/java/ffblockly/meta/AvOptionBuilder.kt
package ffblockly.meta

import ffblockly.meta.AvOption.AvOptionField
import ffblockly.meta.AvOption.AvTarget
import ffblockly.meta.AvOption.AvTarget.*
import java.util.*
import java.util.regex.Matcher

class AvOptionBuilder {
    private var optMatcher: Matcher? = null
    fun setOptMatcher(optMatcher: Matcher?): AvOptionBuilder {
        this.optMatcher = optMatcher
        return this
    }

    fun createAvOptionImpl(): AvOption {
        val optMatcher1 = optMatcher
        return Scan.MYFACTORY.options(object : AvOption {
            override val optionName: String? by lazy { optMatcher1!!.group(AvOptionField.optionName.name) }

            override val type: String? by lazy { optMatcher1!!.group(AvOptionField.type.name) }

            override val targets: List<AvTarget>? by lazy {
                if (type == null) null
                else {
                    values().filter { '.' != optMatcher1!!.group(AvOptionField.targets.name)[it.ordinal] }
                }
            }
            override val description: String? = optMatcher1!!.group(AvOptionField.description.name)

            override val range1: String? = optMatcher1!!.group(AvOptionField.range1.name)

            override val range2: String? = optMatcher1!!.group(AvOptionField.range2.name)

            override val def: String? = optMatcher1!!.group(AvOptionField.def.name)

            override val encoding: Boolean? = type?.let { Encoding in targets!! }?.takeUnless { it == false }

            override val decoding: Boolean? =  type?.let { Decoding in targets!! }?.takeUnless { it == false }

            override val filtering: Boolean? =  type?.let { Filtering in targets!! }?.takeUnless { it == false }

            override val video: Boolean? =  type?.let { Video in targets!! }?.takeUnless { it == false }

            override val audio: Boolean? =  type?.let { Audio in targets!! }?.takeUnless { it == false }

            override val subtitle: Boolean? =  type?.let { Subtitle in targets!! }?.takeUnless { it == false }

            override val export: Boolean? =  type?.let { Export in targets!! }?.takeUnless { it == false }

            override val readonly: Boolean? =  type?.let { Readonly in targets!! }?.takeUnless { it == false }
            override val children: List<AvOption>?
                get() = ch

            protected var ch: List<AvOption>? = null

            override fun addChild(o: AvOption) {
                ch = (this.ch ?: emptyList()) + o
            }
        }).`as`()
    }
}